sec1_intro_and_setup/test_app/.gitignore

```diff
@@ -0,0 +1,40 @@
+# See https://help.github.com/articles/ignoring-files for more about ignoring files.
```

Comment:
1. A `.gitignore` file is used to tell git to exclude files from being added to a given repo. Each line of the file represents a filename, directory, or pattern. When git looks for changes in the working directory, it will automatically ignore any file that matches an entry in the `.gitignore` file.

2. Files should be in `.gitignore`:
- dependency caches
- compiled code
- build output directories, such as /bin, /out, or /target
- files generated at runtime, such as .log, .lock, or .tmp
- hidden system files, such as `.DS_Store` or `Thumbs.db`
- personal IDE config files, such as `.idea/workspace.xml`

3. Files should not be in `.gitignore`:
- Files that are not used by the project
- Files that are not used by anyone in a team
- Files that are generated by another process

Reply:
2. Why all these files should be here?
3. This is not 100% accurate.

Reply:
2. These files should not be committed to the repositories as they can be re-generated. However, these files will still show up whenever you type `git status` if you did not put them in the `.gitignore` file and commit to the repo. Hence, putting these files in `.gitignore` file commands Git to ignore these files. Any file (and directory) that matches won’t show up in `git status`, and if one tries to add them via `git add`, Git will refuse to do so (you can still add them by specifying the -f flag when adding).

3. Files should not be in .gitignore:
- Files that are used by the project
- Files that are used by anyone in a team
- Files that cannot be regenerated

Reply:
The keyword you use here is: `files that can be re-generated`

2. Aside from the statement above, everything is good.
3. Will use this to illustrate the above statement. 

- Files that are used by the project
- Files that are used by anyone in a team
- Files that cannot be regenerated

These are the 3 points you use to determine if a file should be put into `.gitignore`. In this case, why isn't `Gemfile.lock` & `yarn.lock` files here? These 2 files are used by the project, used by anyone & everyone in the team & they are generated with a command.

Reply:
2. Change keyword to:  Things like temporary files, logs, configurations that are specific to a computer, files that are for testing only, private keys for code signing or files that can be easily regenerated should not be commited to your repository.

3.  > A file should be ignored if any of the following is true:
     > The file is not used by your project
     > The file is not used by anyone else in your team
     > The file is generated by another process

     Based on the rules above, 
     `Gemfile.lock` - One should check in the `Gemfile.lock` in applications, and not in gems. The `Gemfile.lock` remembers the exact versions and sources of every piece of third-party code that you use. For applications, you need this. When developing a gem, this can obscure issues that will occur because gems are deployed (using the gem command) without the benefit of bundler.
     `yarn.lock` - One should add `yarn.lock` to your git but do not ignore it.
     > When you run either `yarn` or `yarn add <package>`, Yarn will generate a `yarn.lock` file within the root directory of your package. When other people start using Yarn instead of `npm`, the `yarn.lock` file will ensure that they get precisely the same dependencies as you have.

Reply:
If you read your (2) & (3), there is a discrepency. In (2), you say: `files that can be easily regenerated should not be commited to your repository` but in (3), you say `Gemfile.lock` & `yarn.lock` should be committed to the repo.

`Gemfile.lock` & `yarn.lock` are both easily regenerable files

---

Aside from that, all good